polygeist/tools/cgeist/Lib/CGExpr.cc

```diff
@@ -1133,8 +1133,11 @@ MLIRScanner::emitSYCLOps(const clang::Expr *Expr,
   if (const auto *ConsExpr = dyn_cast<clang::CXXConstructExpr>(Expr)) {
     Func = ConsExpr->getConstructor()->getAsFunction();
 
-    if (mlirclang::isNamespaceSYCL(Func->getEnclosingNamespaceContext()))
-      if (const auto *RD = dyn_cast<clang::CXXRecordDecl>(Func->getParent())) {
+    if (static_cast<bool>(
+            mlirclang::isNamespaceSYCL(Func->getEnclosingNamespaceContext())))
+      if (const auto *RD = dyn_cast<clang::CXXRecordDecl>(Func->getParent());
+          mlirclang::areSYCLMemberFunctionOrConstructorArgs(
+              ValueRange{Args}.getTypes())) {
```

Comment:
First time seeing this, is it only true why RD != nullptr && areSYCLMemberFunctionOrConstructorArgs both true?

Reply:
Yes. I can change the code if you think splitting the conditions might be more readable.

Reply:
mainly want to improve my C++ skills. 😄 

I think is more readable, if `emitSYCLOps` first early exit with `static_cast<bool>(mlirclang::isNamespaceSYCL(Func->getEnclosingNamespaceContext()))`, then this split into 2 ifs.

Reply:
> First time seeing this, is it only true when RD != nullptr && areSYCLMemberFunctionOrConstructorArgs both true?

I don't think there is a nullptr check being done here: https://godbolt.org/z/4nezqYsP9

Reply:
Well, we're checking: `static_cast<bool>(dyn_cast<clang::CXXRecordDecl>(Func->getParent()))` value here to be precise. You're right.

Reply:
The if stmt requires a expression as its condition. `const auto *RD = dyn_cast<clang::CXXRecordDecl>(Func->getParent());` in that context is a stmt, so the expression `mlirclang::areSYCLMemberFunctionOrConstructorArgs(ValueRange{Args}.getTypes())` is always evaluated and its result will be used to determine which branch the `if` takes.

Reply:
This is C++17's init-statement.